### ISSUE TITLE
I'm using lenet training mnist on cpu, how can mxnet makes the most of all the cpus in a machine?

### ISSUE THREAD

**Issue by dushoufu** [dmlc/mxnet#1031]
I only modify the file lenet.py and change   dev = mx.gpu(). to dev = mx.cpu().
I find it runs too slow,and only one cpu is at 100%,others is low.
how can I modify the lenet.py and can use all the cpus?

thank.

**Comment by lukemetz** [comment]
2 things to try:
First, have a look at: http://mxnet.readthedocs.org/en/latest/env_var.html to see if you can bump up number of cpu to use.
Second, I would highly recommend installing openblas and recompiling. Openblas runs its blas on multiple threads as opposed to one.

**Comment by dushoufu** [comment]
Yes，I modify the MXNET_CPU_WORKER_NTHREADS，all the cpu are balanced。 But the speed is not increased obviously，every cpu is not up to 30%。I check the memory and there are enough，and i also ensure i have used Openblas。I don't know what is the bottleneck？

**Comment by dushoufu** [comment]
Is there anything i can check？

**Comment by lukemetz** [comment]
Hmm, that's strange.
How did you install openblas? I know openblas can be compiled without support for this I am not sure if the default binaries disable this as I always build from source. Openblas also has environment variables to enable multiple flags: OPENBLAS_NUM_THREADS. Have a look at https://github.com/xianyi/OpenBLAS.

A quick test that I always do to test blas (assuming you have built numpy to use openblas) is to dot product large random matrices together in a while true. and check cpu usage. This will only help only help debug if numpy is linked against openblas.

**Comment by dushoufu** [comment]
I test numpy using dot and change openblas's environment variables export OPENBLAS_NUM_THREADS=8. But  I find it only use 4 cpus and every is up to 100%. How can it  use all 8 cpus?

**Comment by dushoufu** [comment]
thank，I set  OPENBLAS_NUM，but i find openblas threads is not up to 4，even i make OPENBLAS_NUM >4. In this case,I rerun mxnet,the speed is faster than before with Speed: 554.65 samples/sec,while it is Speed: 410.65 samples/sec before. 
However,mxnet is still a lot slower than caffe in the same machine. caffe only need 8 mins training mnist,while mxnet need 50 mins.

**Comment by dushoufu** [comment]
Does input has so significant impact? The only difference is that caffe uses lmdb as input.

**Comment by lukemetz** [comment]
I doubt that input has a significant impact. The rec loader in mxnet is very fast. I am able to train the GPU example at crazy fast samples/sec so its probably not that.

As for performance, that is quite strange and I am out of ideas. I just ran the example on my cpu and I am able to get 35k ish samples/second with out touching anything on a less good cpu and get 99% accuracy in 18 seconds. (where as on a 980 i finish in 11 seconds at >100k samples a second).

This leads me to think that it is your setup but I am not sure how or what could be causing this. Sorry. Maybe one of the developers of the project has better ideas.

Don't know if this matters, but I built my openblas and numpy from source.

What does your config.mk look like?

**Comment by dushoufu** [comment]
I don't think input has that key role. After all, lmdb is only to be used as input for once only. I run the example directly. For mxnet, i directly run "python lenet.py" without change anything. Following is my
config.mk and lenet.py;
#-------------------------------------------------------------------------------
#  Template configuration for compiling mxnet
#
#  If you want to change the configuration, please use the following
#  steps. Assume you are on the root directory of mxnet. First copy the this
#  file so that any local changes will be ignored by git
#
#  $ cp make/config.mk .
#
#  Next modify the according entries, and then compile by
#
#  $ make
#
#  or build in parallel with 8 threads
#
#  $ make -j8
#-------------------------------------------------------------------------------

#---------------------
# choice of compiler
#--------------------

export CC = gcc
export CXX = g++
export NVCC = nvcc

# whether compile with debug
DEBUG = 0

# the additional link flags you want to add
ADD_LDFLAGS =-L/opt/OpenBLAS/lib

# the additional compile flags you want to add
ADD_CFLAGS =-I/opt/OpenBLAS/include

#---------------------------------------------
# matrix computation libraries for CPU/GPU
#---------------------------------------------

# whether use CUDA during compile
USE_CUDA = 0

# add the path to CUDA libary to link and compile flag
# if you have already add them to enviroment variable, leave it as NONE
# USE_CUDA_PATH = /usr/local/cuda
USE_CUDA_PATH = NONE

# whether use CUDNN R3 library
USE_CUDNN = 0

# whether use opencv during compilation
# you can disable it, however, you will not able to use
# imbin iterator
USE_OPENCV = 1

# use openmp for parallelization
USE_OPENMP = 1

# choose the version of blas you want to use
# can be: mkl, blas, atlas, openblas
USE_STATIC_MKL = NONE
USE_BLAS = openblas

# add path to intel libary, you may need it for MKL, if you did not add the path
# to enviroment variable
USE_INTEL_PATH = NONE

# If use MKL, choose static link automaticly to allow python wrapper
ifeq ($(USE_BLAS), mkl)
	USE_STATIC_MKL = 1
endif

#----------------------------
# distributed computing
#----------------------------

# whether or not to enable mullti-machine supporting
USE_DIST_KVSTORE = 1 

# whether or not allow to read and write HDFS directly. If yes, then hadoop is
# required
USE_HDFS = 0

# path to libjvm.so. required if USE_HDFS=1
LIBJVM=$(JAVA_HOME)/jre/lib/amd64/server

# whether or not allow to read and write AWS S3 directly. If yes, then
# libcurl4-openssl-dev is required, it can be installed on Ubuntu by
# sudo apt-get install -y libcurl4-openssl-dev
USE_S3 = 0


lenet.py
# pylint: skip-file
from data import mnist_iterator
import mxnet as mx
import logging

## define lenet
# input
data = mx.symbol.Variable('data')
# first conv
conv1 = mx.symbol.Convolution(data=data, kernel=(5,5), num_filter=20)
tanh1 = mx.symbol.Activation(data=conv1, act_type="tanh")
pool1 = mx.symbol.Pooling(data=tanh1, pool_type="max",
                          kernel=(2,2), stride=(2,2))
# second conv
conv2 = mx.symbol.Convolution(data=pool1, kernel=(5,5), num_filter=50)
tanh2 = mx.symbol.Activation(data=conv2, act_type="tanh")
pool2 = mx.symbol.Pooling(data=tanh2, pool_type="max",
                          kernel=(2,2), stride=(2,2))
# first fullc
flatten = mx.symbol.Flatten(data=pool2)
fc1 = mx.symbol.FullyConnected(data=flatten, num_hidden=500)
tanh3 = mx.symbol.Activation(data=fc1, act_type="tanh")
# second fullc
fc2 = mx.symbol.FullyConnected(data=tanh3, num_hidden=10)
# loss
lenet = mx.symbol.SoftmaxOutput(data=fc2, name='softmax')

## data
train, val = mnist_iterator(batch_size=100, input_shape=(1,28,28))

## train
logging.basicConfig(level=logging.DEBUG)
# dev = [mx.gpu(i) for i in range(2)]
#dev = mx.gpu()
dev = mx.cpu()
model = mx.model.FeedForward(
    ctx = dev, symbol = lenet, num_epoch = 20,
    learning_rate = 0.05, momentum = 0.9, wd = 0.00001)
model.fit(X=train, eval_data=val,
          batch_end_callback=mx.callback.Speedometer(100))

**Comment by dushoufu** [comment]
Compared to original file, I change config.mk use USE_DIST_KVSTORE = 1 ;

**Comment by Adesun** [comment]
If your input is not binary file, you can check out your disk IO load status when training.

**Comment by dushoufu** [comment]
io wait 0,so io is not the bottleneck.

**Comment by dushoufu** [comment]
I set OPENBLAS_NUM > 4，but i find openblas threads is actually not up to 4，even though i make OPENBLAS_NUM >4. I use numpy to test and the result is that only 4 cpu cores is up to 100%. If I can make all the cpu up to 100%, mxnet probably can be faster. 
But correspondingly, caffe which use the same openblas in the same machine can also be faster. The speed gap between caffe and mxnet is still significant.

**Comment by antinucleon** [comment]
Thanks for letting us know. I will check CPU version performance. Most time we only pay attention to GPU speed so maybe there is some bug.

**Comment by dushoufu** [comment]
I’m doing a project which intend to use distributed version of caffe or mxnet. We have a lot of cpus,so we hope to make a choice between caffe and mxnet,soo I do this test.  Normally, I think if we want to do a   real project for commercial target, the performance of cpu need to be considered,as most of us have many cpus instead of gpus if we use distributed deep learning system.

**Comment by dushoufu** [comment]
I’m doing a project which intend to use distributed version of caffe or mxnet. We have a lot of cpus,so we hope to make a choice between caffe and mxnet,so I do this test.  Normally, I think if we want to do a   real project for commercial target, the performance of cpu need to be considered,as most of us have many cpus instead of gpus if we use distributed deep learning system.

**Comment by junshui** [comment]
I observed the same problem. 

I have a computer with 12 cores from 2 CPUs. When I run mxnet, the CPU usage is only about 10%, which does not change no matter how I setup MXNET_CPU_WORKER_NTHREADS and OPENBLAS_NUM_THREADS.
